From 223d96afe79925be37cb8e36e5f5d50eb716eed1 Mon Sep 17 00:00:00 2001
From: Hien Huynh <hien.huynh.px@renesas.com>
Date: Fri, 6 Mar 2020 10:35:04 +0700
Subject: [PATCH 187/197] arm64: dts: renesas: r8a774c0-{,es10}-cat874: Enable
 Audio support in Low speed connector

This patch enable sound support in Low speed connector
for the CAT874 board on the R8A774C0 SoC by adding
device node of the PCM5102 codec (External I2S Board
in Low speed Connector).

Therefore, we support 2 audio codecs in CAT874 plarform:
1. HDMI audio from TDA19988 HDMI transmitter. (default)
2. External I2S Audio Codec PCM5102A in Low Speed Connector.

Signed-off-by: Hien Huynh <hien.huynh.px@renesas.com>
---
 arch/arm64/boot/dts/renesas/r8a774c0-cat874.dts    | 56 ++++++++++++++++++----
 .../boot/dts/renesas/r8a774c0-es10-cat874.dts      | 56 ++++++++++++++++++----
 2 files changed, 92 insertions(+), 20 deletions(-)

diff --git a/arch/arm64/boot/dts/renesas/r8a774c0-cat874.dts b/arch/arm64/boot/dts/renesas/r8a774c0-cat874.dts
index a6f7559..b1579d0 100644
--- a/arch/arm64/boot/dts/renesas/r8a774c0-cat874.dts
+++ b/arch/arm64/boot/dts/renesas/r8a774c0-cat874.dts
@@ -114,20 +114,46 @@
 	sound: sound {
 		compatible = "simple-audio-card";
 
-		simple-audio-card,name = "CAT874 HDMI sound";
 		simple-audio-card,format = "i2s";
-		simple-audio-card,bitclock-master = <&sndcpu>;
-		simple-audio-card,frame-master = <&sndcpu>;
+		simple-audio-card,name = "rcar_sound";
 
-		sndcodec: simple-audio-card,codec {
-			sound-dai = <&tda19988>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		simple-audio-card,dai-link@0 {
+			reg = <0>;
+			format = "i2s";
+			cpu {
+				sound-dai = <&rcar_sound 0>;
+				bitclock-master;
+				frame-master;
+			};
+
+			codec {
+				sound-dai = <&tda19988>;
+			};
 		};
 
-		sndcpu: simple-audio-card,cpu {
-			sound-dai = <&rcar_sound>;
+		simple-audio-card,dai-link@1 {
+			reg = <1>;
+			format = "i2s";
+			cpu {
+				sound-dai = <&rcar_sound 1>;
+				bitclock-master;
+				frame-master;
+			};
+
+			codec {
+				sound-dai = <&pcm5102a>;
+			};
 		};
 	};
 
+	pcm5102a: pcm5102a {
+		compatible = "ti,pcm5102a";
+		#sound-dai-cells = <0>;
+	};
+
 	vcc_sdhi0: regulator-vcc-sdhi0 {
 		compatible = "regulator-fixed";
 
@@ -407,12 +433,14 @@
 	};
 
 	sound_clk_pins: sound_clk {
-		groups = "audio_clkout1_a";
+		groups = "audio_clkout1_a",
+			 "audio_clk_a", "audio_clk_b_a", "audio_clk_c_a";
 		function = "audio_clk";
 	};
 
 	sound_pins: sound {
-		groups = "ssi01239_ctrl", "ssi0_data";
+		groups = "ssi01239_ctrl", "ssi0_data",
+			 "ssi349_ctrl", "ssi3_data", "ssi4_data";
 		function = "ssi";
 	};
 
@@ -448,7 +476,7 @@
 	pinctrl-names = "default";
 
 	/* Single DAI */
-	#sound-dai-cells = <0>;
+	#sound-dai-cells = <1>;
 
 	/* audio_clkout0/1/2/3 */
 	#clock-cells = <1>;
@@ -460,9 +488,17 @@
 		dai0 {
 			playback = <&ssi0 &src0 &dvc0>;
 		};
+
+		dai1 {
+			playback = <&ssi4 &src4 &dvc1>;
+		};
 	};
 };
 
+&ssi4 {
+	shared-pin;
+};
+
 &rwdt {
 	timeout-sec = <60>;
 	status = "okay";
diff --git a/arch/arm64/boot/dts/renesas/r8a774c0-es10-cat874.dts b/arch/arm64/boot/dts/renesas/r8a774c0-es10-cat874.dts
index fd29a3f..c8c778f 100644
--- a/arch/arm64/boot/dts/renesas/r8a774c0-es10-cat874.dts
+++ b/arch/arm64/boot/dts/renesas/r8a774c0-es10-cat874.dts
@@ -114,20 +114,46 @@
 	sound: sound {
 		compatible = "simple-audio-card";
 
-		simple-audio-card,name = "CAT874 HDMI sound";
 		simple-audio-card,format = "i2s";
-		simple-audio-card,bitclock-master = <&sndcpu>;
-		simple-audio-card,frame-master = <&sndcpu>;
+		simple-audio-card,name = "rcar_sound";
 
-		sndcodec: simple-audio-card,codec {
-			sound-dai = <&tda19988>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		simple-audio-card,dai-link@0 {
+			reg = <0>;
+			format = "i2s";
+			cpu {
+				sound-dai = <&rcar_sound 0>;
+				bitclock-master;
+				frame-master;
+			};
+
+			codec {
+				sound-dai = <&tda19988>;
+			};
 		};
 
-		sndcpu: simple-audio-card,cpu {
-			sound-dai = <&rcar_sound>;
+		simple-audio-card,dai-link@1 {
+			reg = <1>;
+			format = "i2s";
+			cpu {
+				sound-dai = <&rcar_sound 1>;
+				bitclock-master;
+				frame-master;
+			};
+
+			codec {
+				sound-dai = <&pcm5102a>;
+			};
 		};
 	};
 
+	pcm5102a: pcm5102a {
+		compatible = "ti,pcm5102a";
+		#sound-dai-cells = <0>;
+	};
+
 	vcc_sdhi0: regulator-vcc-sdhi0 {
 		compatible = "regulator-fixed";
 
@@ -407,12 +433,14 @@
 	};
 
 	sound_clk_pins: sound_clk {
-		groups = "audio_clkout1_a";
+		groups = "audio_clkout1_a",
+			 "audio_clk_a", "audio_clk_b_a", "audio_clk_c_a";
 		function = "audio_clk";
 	};
 
 	sound_pins: sound {
-		groups = "ssi01239_ctrl", "ssi0_data";
+		groups = "ssi01239_ctrl", "ssi0_data",
+			 "ssi349_ctrl", "ssi3_data", "ssi4_data";
 		function = "ssi";
 	};
 
@@ -448,7 +476,7 @@
 	pinctrl-names = "default";
 
 	/* Single DAI */
-	#sound-dai-cells = <0>;
+	#sound-dai-cells = <1>;
 
 	/* audio_clkout0/1/2/3 */
 	#clock-cells = <1>;
@@ -460,9 +488,17 @@
 		dai0 {
 			playback = <&ssi0 &src0 &dvc0>;
 		};
+
+		dai1 {
+			playback = <&ssi4 &src4 &dvc1>;
+		};
 	};
 };
 
+&ssi4 {
+	shared-pin;
+};
+
 &rwdt {
 	timeout-sec = <60>;
 	status = "okay";
-- 
2.7.4

